dodiscover/constraint/pcalg.py

```diff
@@ -0,0 +1,309 @@
+    max_combinations : int, optional
+        Maximum number of tries with a conditioning set chosen from the set of possible
+        parents still, by default None. If None, then will not be used. If set, then
+        the conditioning set will be chosen lexographically based on the sorted
+        test statistic values of 'ith Pa(X) -> X', for each possible parent node of 'X'.
+    apply_orientations : bool
+        Whether or not to apply orientation rules given the learned skeleton graph
+        and separating set per pair of variables. If ``True`` (default), will
+        apply Meek's orientation rules R0-3, orienting colliders and certain
+        arrowheads :footcite:`Meek1995`.
+    ci_estimator_kwargs : dict
+        Keyword arguments for the ``ci_estimator`` function.
+
+    Attributes
+    ----------
+    graph_ : EquivalenceClassProtocol
+        The equivalence class of graphs discovered.
+    separating_sets_ : dict of dict of list of set
+        The dictionary of separating sets, where it is a nested dictionary from
+        the variable name to the variable it is being compared to the set of
+        variables in the graph that separate the two.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    graph_: EquivalenceClassProtocol
+    separating_sets_: Optional[Dict[Column, Dict[Column, List[Set[Column]]]]]
+
+    def __init__(
+        self,
+        ci_estimator: BaseConditionalIndependenceTest,
+        alpha: float = 0.05,
+        min_cond_set_size: int = None,
+        max_cond_set_size: int = None,
+        max_iter: int = 1000,
+        max_combinations: int = None,
+        apply_orientations: bool = True,
+        **ci_estimator_kwargs,
+    ):
+        super().__init__(
+            ci_estimator,
+            alpha,
+            min_cond_set_size=min_cond_set_size,
+            max_cond_set_size=max_cond_set_size,
+            max_combinations=max_combinations,
+            **ci_estimator_kwargs,
+        )
+        self.max_iter = max_iter
+        self.apply_orientations = apply_orientations
+
+    def convert_skeleton_graph(self, graph: nx.Graph) -> EquivalenceClassProtocol:
+        """Convert skeleton graph as undirected networkx Graph to CPDAG.
+
+        Parameters
+        ----------
+        graph : nx.Graph
+            Converts a skeleton graph to the representation needed
+            for PC algorithm, a CPDAG.
+
+        Returns
+        -------
+        graph : EquivalenceClassProtocol
+            The CPDAG class.
+        """
+        from pywhy_graphs import CPDAG
+
+        # convert Graph object to a CPDAG object with
+        # all undirected edges
+        graph = CPDAG(incoming_undirected_edges=graph)
+        return graph
+
+    def orient_edges(self, graph: EquivalenceClassProtocol) -> None:
+        """Orient edges in a skeleton graph to estimate the causal DAG, or CPDAG.
+
+        These are known as the Meek rules :footcite:`Meek1995`. They are deterministic
```

Comment:
How would these work with required edges and excluded edges?  Does it make sense to build in those constraints now?  Those are the only real domain knowledge constraints we can provide for constraint-based models.